Mathlib/Algebra/MvPolynomial/Funext.lean

```diff
@@ -24,40 +24,54 @@ if they are equal upon evaluating them on an arbitrary assignment of the variabl
 
 namespace MvPolynomial
 
-variable {R : Type*} [CommRing R] [IsDomain R] [Infinite R]
+variable {R : Type*} [CommRing R] [IsDomain R]
 
 private theorem funext_fin {n : ℕ} {p : MvPolynomial (Fin n) R}
-    (h : ∀ x : Fin n → R, eval x p = 0) : p = 0 := by
+    (s : Fin n → Set R) (hs : ∀ i, (s i).Infinite)
+    (h : ∀ x ∈ Set.pi .univ s, eval x p = 0) : p = 0 := by
   induction n with
   | zero =>
     apply (MvPolynomial.isEmptyRingEquiv R (Fin 0)).injective
     rw [RingEquiv.map_zero]
-    convert h finZeroElim
+    convert h _ finZeroElim
   | succ n ih =>
     apply (finSuccEquiv R n).injective
-    simp only [map_zero]
-    refine Polynomial.funext fun q => ?_
-    rw [Polynomial.eval_zero]
-    apply ih fun x => ?_
-    calc _ = _ := eval_polynomial_eval_finSuccEquiv p _
-         _ = 0 := h _
+    rw [map_zero]
+    apply Polynomial.eq_zero_of_infinite_isRoot
+    apply ((hs 0).image (C_injective ..).injOn).mono
+    rintro _ ⟨r, hr, rfl⟩
+    refine ih (s ·.succ) (fun _ ↦ hs _) fun x hx ↦ ?_
+    rw [eval_polynomial_eval_finSuccEquiv]
+    exact h _ fun i _ ↦ i.cases (by simpa using hr) (by simpa using hx)
 
-/-- Two multivariate polynomials over an infinite integral domain are equal
-if they are equal upon evaluating them on an arbitrary assignment of the variables. -/
-theorem funext {σ : Type*} {p q : MvPolynomial σ R} (h : ∀ x : σ → R, eval x p = eval x q) :
+variable {σ : Type*} {p q : MvPolynomial σ R}
+
+theorem funext_set (s : σ → Set R) (hs : ∀ i, (s i).Infinite)
```

Comment:
Could you add an `iff` version, and a docstring mentioning the "infinite box" idea?

Reply:
The backwards statement is probably not that interesting as you can just `subst` to get it by `rfl`, and you need to specify `s` (which is not the case for `funext`).

Reply:
added `funext_set_iff`